public/app/plugins/datasource/testdata/nodeGraphUtils.ts

```diff
@@ -0,0 +1,160 @@
+import { nodes, edges } from './testData/serviceMapResponse';
+
+export function generateRandomNodes(count = 10) {
+  const nodes = [];
+
+  const root = {
+    id: '0',
+    title: 'root',
+    subTitle: 'client',
+    success: 1,
+    error: 0,
+    stat1: Math.random(),
+    stat2: Math.random(),
+    edges: [] as any[],
+  };
+  nodes.push(root);
+  const nodesWithoutMaxEdges = [root];
+
+  const maxEdges = 3;
+
+  for (let i = 1; i < count; i++) {
+    const node = makeRandomNode(i);
+    nodes.push(node);
+    const sourceIndex = Math.floor(Math.random() * Math.floor(nodesWithoutMaxEdges.length - 1));
+    const source = nodesWithoutMaxEdges[sourceIndex];
+    source.edges.push(node.id);
+    if (source.edges.length >= maxEdges) {
+      nodesWithoutMaxEdges.splice(sourceIndex, 1);
+    }
+    nodesWithoutMaxEdges.push(node);
+  }
+
+  // Add some random edges to create possible cycle
+  const additionalEdges = Math.floor(count / 2);
+  for (let i = 0; i <= additionalEdges; i++) {
+    const sourceIndex = Math.floor(Math.random() * Math.floor(nodes.length - 1));
+    const targetIndex = Math.floor(Math.random() * Math.floor(nodes.length - 1));
+    if (sourceIndex === targetIndex || nodes[sourceIndex].id === '0' || nodes[sourceIndex].id === '0') {
+      continue;
+    }
+
+    nodes[sourceIndex].edges.push(nodes[sourceIndex].id);
+  }
+
+  const nodeFields: any = {
+    id: {
+      values: new ArrayVector(),
+      type: FieldType.string,
+    },
+    title: {
+      values: new ArrayVector(),
+      type: FieldType.string,
+      labels: { NodeGraphValueType: 'title' },
+    },
+    subTitle: {
+      values: new ArrayVector(),
+      type: FieldType.string,
+      labels: { NodeGraphValueType: 'subTitle' },
+    },
+    stat1: {
+      values: new ArrayVector(),
+      type: FieldType.number,
+      labels: { NodeGraphValueType: 'mainStat' },
+    },
+    stat2: {
+      values: new ArrayVector(),
+      type: FieldType.number,
+      labels: { NodeGraphValueType: 'secondaryStat' },
+    },
+    success: {
+      values: new ArrayVector(),
+      type: FieldType.number,
+      labels: { NodeGraphValueType: 'arc' },
+      config: { color: { fixedColor: 'green' } },
+    },
+    error: {
+      values: new ArrayVector(),
+      type: FieldType.number,
+      labels: { NodeGraphValueType: 'arc' },
+      config: { color: { fixedColor: 'red' } },
+    },
+  };
+
+  const nodeFrame = new MutableDataFrame({
+    name: 'nodes',
+    fields: Object.keys(nodeFields).map(key => ({
+      ...nodeFields[key],
+      name: key,
+    })),
+    meta: { preferredVisualisationType: 'nodeGraph' },
+  });
+
+  const edgeFields: any = {
+    id: {
+      values: new ArrayVector(),
+      type: FieldType.string,
+    },
+    source: {
+      values: new ArrayVector(),
+      type: FieldType.string,
+    },
+    target: {
+      values: new ArrayVector(),
+      type: FieldType.string,
+    },
+  };
```

Comment:
This data structure with 2 data frames and a bunch of hard code labels makes this visualization very hard for any other data source to use with out a bunch of specific code. 

Is this dual data frame structure needed ? have you explored a single data frame structure like this? 

| Source | Target | Title                       | SubTitle       | MainStat | SecStat | ErrorRate |
| ------ | ------ | --------------------------- | -------------- | -------- | ------- | --------- |
| Node1  | Node2  | NetworkA                    | Some sub title | 10       | 20      | 0.2       |
| Node2  | Node3  | NetworkB                    | Another sub    | 10       | 20      | 0.2       |
| Node3  |        | ServerA (edge node)         | sub title      | 20       | 10      | 0.1       |
| Node2  | Node4  |                             |                | 25       | 1       | 0.5       |
| Node4  |        | ServerB (another edge node) |                | 50       | 20      | 0.3       |

Where each node is a row and connections are defined with the target field. To create one to many connections the same Source property would be duplicate (but with different targets).